### ISSUE TITLE
block hook called twice if uc_mem_protect is used inside hook

### ISSUE THREAD

**Issue by liangjs** [unicorn-engine/unicorn#1222]
```c
#include <unicorn/unicorn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
   0:	90                 	nop
   1:	eb 00              	jmp    3
   3:	90                 	nop
 */
#define X86_CODE "\x90\xeb\x00\x90"

#define ADDRESS1 0x1000
#define ADDRESS2 0x2000

static void hook_block(uc_engine *uc, uint64_t address, uint32_t size, void *user_data)
{
	printf(">>> Tracing basic block at 0x%lx, block size = 0x%x\n", address, size);
	if (address == ADDRESS1)
		uc_mem_protect(uc, ADDRESS2, 0x1000, UC_PROT_READ);
}

int main(int argc, char **argv, char **envp)
{
	uc_engine *uc;
	uc_err err;

	err = uc_open(UC_ARCH_X86, UC_MODE_64, &uc);
	if (err != UC_ERR_OK) {
		printf("Failed on uc_open() with error returned: %u\n", err);
		return -1;
	}

	uc_mem_map(uc, ADDRESS1, 0x1000, UC_PROT_ALL);
	uc_mem_map(uc, ADDRESS2, 0x1000, UC_PROT_ALL);

	uc_mem_write(uc, ADDRESS1, X86_CODE, sizeof(X86_CODE) - 1);

	uc_hook trace_hook;
	uc_hook_add(uc, &trace_hook, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

	err = uc_emu_start(uc, ADDRESS1, ADDRESS1 + sizeof(X86_CODE) - 1, 0, 0);
	if (err) {
		printf("Failed on uc_emu_start() with error returned %u: %s\n",
		       err, uc_strerror(err));
	}

	printf("Emulation done.\n");

	uc_close(uc);

	return 0;
}
```

The output is 
```
>>> Tracing basic block at 0x1000, block size = 0x3
>>> Tracing basic block at 0x1000, block size = 0x3
>>> Tracing basic block at 0x1003, block size = 0x1
Emulation done.
```

But it is expected to be
```
>>> Tracing basic block at 0x1000, block size = 0x3
>>> Tracing basic block at 0x1003, block size = 0x1
Emulation done.
```

**Comment by gerph** [comment]
I believe this to be an instance of #1193 which is fixed in PR #1194.

**Comment by gerph** [comment]
Having said that, the change of the memory mapping might actually be a _valid_ case of the block having to be invalidated and therefore rebuilt and thus re-built and re-entered.

**Comment by liangjs** [comment]
I think this is a different issue. It is not fixed in PR #1194. 
In my case, `tcg_exit_req` is cleared, but `uc_map_protect` sets a `uc->quit_request` flag, which results in uc running the block again.
Is this a valid case, even we just change permission of a memory range unrelated to current block?

**Comment by gerph** [comment]
Sorry for the misdirect to that issue - it seemed related on first glance.
I'm not sure whether it's a valid case for unrelated changes to the memory protection though.

My understanding of that TCG trigger is that if you cause a state change it doesn't know that the block is valid any more, so for safety the memory is regenerated. In the MMU case, it could have been the case that you changed the mapping of the location that is being executed, so it's likely that in that case you should trigger a rebuild of the block.

Identifying when the change triggered by an operation in a hook affects a currently executing block is probably tricky - I've not looked at that code at all, but I suspect that not only does the change in protection you're doing not know the current execution (there might not be a current execution - we might not be in a hook), keeping them separated is desirable because of the complexity it would introduce to the memory remapping (and potentially the register state, if we're also thinking of things like the control registers for the MMU being enabled).

You could try updating the code that does the quit request and update it to check for uc->current_cpu to check the actual state of the execution and whether you believe it might be affected by the change in protection. It'd by a little invasive, but might help your case?

Hopefully the author has a better answer for what the intent is for mutable (but unrelated!) state here.



ASIDE, in case it's useful: When I was testing #1194, I also added in the code hooks to convince myself that nothing was actually executed. I suspect that if you changed the hook from triggering on the block for `if (address == ADDRESS1)` but by a code hook for `ADDRESS1 + 1`, you'd find the you'd get a sequence something like:

* block hook for ADDRESS1, size 3
* code hook for ADDRESS1
* code hook for ADDRESS1+1
   * you modify the protect
* block hook for ADDRESS1+1, size 2
* code hook for ADDRESS1+1
   * you modify the protect, but this time it's a no-op
* exits because we reached the terminal address

That really only helps in understanding what's going on; not the actual behavior that you might need, I guess.

**Comment by liangjs** [comment]
I agree that block should be rebuilt and code should only be executed once. But I still think the hook should only be called once.
By the way, the hook is called twice even if we try to stop the emulation (together with `uc_mem_protect`) inside the hook. Like this:
```c
static void hook_block(uc_engine *uc, uint64_t address, uint32_t size, void *user_data)
{
	printf(">>> Tracing basic block at 0x%lx, block size = 0x%x\n", address, size);
	if (address == ADDRESS1) {
		uc_mem_protect(uc, ADDRESS2, 0x1000, UC_PROT_READ);
		uc_emu_stop(uc);
	}
}
/* Output becomes:
>>> Tracing basic block at 0x1000, block size = 0x3
>>> Tracing basic block at 0x1000, block size = 0x3
Emulation done.
*/
```

**Comment by gerph** [comment]
I agree that it'd be nice to only called once.

The emu_stop is genuinely surprising though; I'd have thought it'd have left the block invalidated and just exited. That might be an independant issue.

**Comment by wtdcode** [comment]
Link to #1449

**Comment by wtdcode** [comment]
Fixed in 345b63ee96e82b7c98828b6efbcf87b1eff33e13 and unit test is added.
